### PR TITLE
Autogenerate and print elastic pwd on startup

### PR DESCRIPTION
**Pull request by jkakavas**
Introduces functionality to generate and set a password for the
elastic user during the initialization of the Security plugin if

- `bootstrap.password` is not already set in the keystore
- the security index doesn't already exist
- and the password for the elastic user is not yet set ( the doc
for the user doesn't exist in the security index )

This is printed in INFO level in the log, as we have already
wrapped System.out in Log4j and this is where stdout would be
printed either way. The password should not make it to the logs
and in a follow up we need to come up with a solution to this
problem.

### CODE REVIEW

**Comment by jkakavas** [comment]
This is missing package tests, I plan to add some after/when we merge #77231, if we decide this is worth merging in its current format.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (jkakavas) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:650]: We could pass a reference to the original PrintStream from Bootstrap to Node and then to the SecurityPlugin's `createComponents` so that we can temporarily call `System.setOut` to that , print to stdout and revert to log4j afterwards. This needs `setIO` RuntimePermission for the security plugin though and I'm not sure how comfortable I am with this change.

**Comment by tvernum** [comment]
> generate and set a password for the elastic user during the initialization of the Security plugin if
> - bootstrap.password is not already set in the keystore
> - the security index doesn't already exist 
> - and the password for the elastic user is not yet set ( the doc for the user doesn't exist in the security index )

Do we intend, at a future time for there to be a setting that disables auto-configuration?

It feels like there might be use cases where a user wishes to not have a superuser password printed to the comsole, and we're currently saying the only way to achieve that is to set a `bootstrap.password`.
I think a way to opt-out of auto-security-configuration like this is reasonable.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (tvernum) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:650]: I don't understand why you'd need to call `System.setOut` - you just need a reference to the `PrintStream` so you can use it.

What we need is for `Bootstrap` (or something close to it) to hold on to the original stdout in a place that we can get to it (but is on the forbidden APIs list).

`Bootstrap` already has code to close original stream if we're not in the foreground, so I think a change to add a `SystemConsole` class in `org.elasticsearch.bootstrap` is entirely feasible.

Do you want me to pull something together?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:639]: I'd prefer this wasn't in `Security`.
We have a bad habit of just adding more and more into this class, and it's unnecessary. 

Can we have a `GenerateInitElasticPasswordListener`?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:502]: If this is a couple of lines lower, then we can remove the `SetOnce` entirely and just pass the `NativeUsersStore` to the listener object's constructor (if we replace the method reference with a class).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:313]: Is this 90% identical to the `updateReservedUser` method?

The main difference is just the `opType`, and it doesn't feel like we need to duplicate all the code for that.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:309]: Are we sure we never need to clear the realm cache at this point?

Given this is a public method, with no indication that it should only be called at node startup, I think we should clear the cache.
If a reserved user is created _after_ someone has tried to authenticate with it, a node might have a negative cached password.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:291]: We lack any tests for this method.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:668]: I think the listener (if it's a real object) should remove itself at this point. We don't need to call it ever again.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:650]: > I don't understand why you'd need to call System.setOut - you just need a reference to the PrintStream so you can use it.

Yes, you are right. I misunderstood how we do it ( and how handling stdout works most probably ) in LogConfigurator, and I had presumed the `PrintStream` would be unusable. 

> Do you want me to pull something together?

That would be great

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:639]: Sure thing, I will amend

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:309]: Good point

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:313]: Yes. Late night interface design is the worst. I'll add the opType as a parameter since also https://github.com/elastic/elasticsearch/pull/77291/files#r702622693 makes sense

**Comment by jkakavas** [comment]
> > generate and set a password for the elastic user during the initialization of the Security plugin if
> > 
> > * bootstrap.password is not already set in the keystore
> > * the security index doesn't already exist
> > * and the password for the elastic user is not yet set ( the doc for the user doesn't exist in the security index )
> 
> Do we intend, at a future time for there to be a setting that disables auto-configuration?
> 
> It feels like there might be use cases where a user wishes to not have a superuser password printed to the comsole, and we're currently saying the only way to achieve that is to set a `bootstrap.password`.
> I think a way to opt-out of auto-security-configuration like this is reasonable.

Yes, there is a relevant discussion in https://github.com/elastic/elasticsearch/pull/77231#discussion_r701847346. Even know we could do it with enrollment mode setting but I plan on adding the new setting in #77231

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: Done in 79a6c21

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:310]: 79a6c21

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java:68]:
```suggestion
                                    LOGGER.info("Please note these down as they will not be shown again.");
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java:46]: Why a `GroupedActionListener` when you're handling the sequencing yourself?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:396]:
```suggestion

```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:262]:
```suggestion
                                    updateReservedUser(
                                        username,
                                        request.passwordHash(),
                                        DocWriteRequest.OpType.INDEX,
                                        request.getRefreshPolicy(),
                                        listener
                                    );
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java:43]: I see you added `removeStateListener` to the IndexManager but don't actually call it here.

**Review by tvernum — APPROVED** [review]
I left a few comments, but I'm happy for you to merge when you and @BigPandaToo are happy.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java
Comment (tvernum) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java:43]:
```suggestion

            securityIndexManager.removeStateListener(this);

```

Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java:43]: got lost when I added the kibana_system part, will add

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/GenerateInitialBuiltinUsersPasswordListener.java]: Yes, it was actually unnecessary. I think I just misread the docs and got an incorrect impression that I could fire off the two async actions, wait for both to complete and then do the outputting. Thinking through this again, I could probably do that with a CountDownLatch, but I don't think there is anything inherently wrong with sequencing the two calls and output to terminal in the `onResponse` of the second

**Review by BigPandaToo — APPROVED** [review]
LGTM

**Comment by jkakavas** [comment]
I'll wait for a few hours in case we can merge https://github.com/elastic/elasticsearch/pull/77299 which we can take advantage of in order to print to stdout only ( not to logs )  which is the desired end goal for this change

**Comment by rjernst** [comment]
> I'll wait for a few hours in case we can merge #77299

I left a suggestion in #77299 to use `System.console()` directly. I believe this will allow the desired behavior, to print only when in an interactive session, without needing any changes to the existing bootstrap code.

**Comment by jkakavas** [comment]
> I left a suggestion in #77299 to use System.console() directly. I believe this will allow the desired behavior, to print only when in an interactive session, without needing any changes to the existing bootstrap code.

Sounds like a good idea. I'll give it a try here, do some manual testing and see if Packaging Tests are happy too.

**Comment by jkakavas** [comment]
I'll refrain from merging as-is for now just in case I am missing something obvious regarding https://github.com/elastic/elasticsearch/pull/77291#issuecomment-914491853 and will revisit early in my morning in case there are any updates in #77299 and/or relevant discussion @tvernum @rjernst

**Comment by rjernst** [comment]
> Seems like System.console() returns null always

What environment was ES running in when it returned null? It will *only* return non-null if attached to an actual console.

**Comment by tvernum** [comment]
I've confirmed this is due to the way we pass the keystore password via stdin - it's no longer attached to a terminal, so the Console is null 
https://github.com/elastic/elasticsearch/pull/77299#issuecomment-914683364

**Pull request merged**
